net/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+//! holochain_net is a library that defines an abstract networking layer for
+//! different network transports and implements a PeerStore for mapping and
+//! managing the topology of transport layers with regard to relay's e.g. for NAT
+
+#[macro_use]
+extern crate failure;
+
+use failure::Error;
+
+pub mod error;
+
+pub type SerializedAddress = Vec<u8>;
+pub type TransportAddress = Vec<u8>;
+pub type SerializedMessage = Vec<u8>;
+
+/// this closure type will get called when the send completes and the parameter will be the response message (or error)
+type SendResponseClosure = Box<FnMut(Result<SerializedMessage, Error>) -> Option<Error> + Send>;
+
+/// this closure type gets called when a new message arrives, you can respond with a Message or an error.
+type ReceiveClosure =
+    Box<FnMut(&SerializedAddress, &SerializedMessage) -> Result<SerializedMessage, Error> + Send>;
+
+/// trait to encapsulate a node that represents a given Holochain app from the Transport's perspective
+pub trait Node {
+    fn get_address(&self) -> &SerializedAddress;
+    fn get_transport_address(&self) -> TransportAddress;
+}
+
+pub trait Transport {
+    /// initialize the transport
+    /// this might be used for example in a TCP based transport to specify a listening port for
+    /// for the transport.
+    fn initialize(&mut self, config: Option<String>) -> Result<(), Error>;
+
+    /// return a default config with best configuration guesses for the transport
+    fn get_default_config(&self) -> String;
+
+    /// register a peer as a node in the transport
+    fn new_node(
+        &mut self,
+        addr: SerializedAddress,
+        handler: Option<ReceiveClosure>,
+    ) -> Result<(), Error>;
```

Comment:
Per our lucidchart discussion, should all of the SerializedAddress references be moved to a layer above this? I.e. instead of `new_node` we would have:

```rust
fn get_transport_address(&self) -> TransportAddress;
```

1. Then we would layer a DHT per app on top of the strict p2p transport layer (that is, each app would have a unique SerializedAddress, and use that address to discover / track the opaque p2p TransportAddresses of neighbors and index DHT buckets)
2. If a specific app wanted to message another agent within that same app, a higher level abstraction would allow them to just specify a SerializedAddress. It would then make a high-level DHT query to discover the opaque TransportAddress that app agent is currently connected to, bundle up a layered message like:

```javascript
{
  "to": <TransportAddress>,
  "message": {
    "to": <SerializedAddress>,
    "message": { /* ... */ }
  }
}
```

Now that I've written all this, perhaps THIS is the higher level abstraction, and what we really need to do is define the lower-level one?